### ISSUE TITLE
Bad L parameter

### ISSUE THREAD

**Issue by Guigue79** [dmitryd/typo3-realurl#549]
Hi
When users or bots try to set Language variable to fr
I have this error
Thu, 12 Oct 2017 09:46:25 +0200 [ERROR] request="eb2a025b3e3ec" component="DmitryDulepov.Realurl.Encoder.UrlEncoder": Bad "L" parameter ("fr") was detected by realurl......
Logs are very very very large and saturate my disk
I try to ignore this param with config.linkVars = L(0-1) but it does not work

**Comment by osahner** [comment]
Same here. Logging seems to be messed up this release. One Line has more than 3mb!

**Comment by responseinformationsdesign** [comment]
RealURL also seems to ignore [SYS][systemLogLevel].

Errors get written in TYPO3-Log even [SYS][systemLogLevel] was set to only log fatal errors.

**Comment by hannesbochmann** [comment]
@responseinformationsdesign 
realurl uses the logging system ($TYPO3_CONF_VARS['LOG']) of TYPO3 not the simple system log. System log level has no effect on the logging system. You would need to remove the default FileWriter for log level WARNING after you copied it to the log configuration for log level ERROR. This way you should only have errors in typo3temp/logs/typo3.log. By the way this is no recommended. Warnings shouldn't be ignored. ;-)

**Comment by bavarianbytes** [comment]
Same here

**Comment by fazzyx** [comment]
We have the same situation after upgrading realurl to 2.3. 
TYPO3 7.6

**Comment by Batman777** [comment]
Typo3 - 6.31
Same problem with huge log files. Will there be a fix in the near future?

> You would need to remove the default FileWriter for log level WARNING after you copied it to the log configuration for log level ERROR.

I did not get that. Could you please try in other words?

**Comment by hannesbochmann** [comment]
@Batman777 
Put that in your AdditionalConfiguration.php:
```
$GLOBALS['TYPO3_CONF_VARS']['LOG']['writerConfiguration'][\TYPO3\CMS\Core\Log\LogLevel::ERROR]['TYPO3\\CMS\\Core\\Log\\Writer\\FileWriter'] = $GLOBALS['TYPO3_CONF_VARS']['LOG']['writerConfiguration'][\TYPO3\CMS\Core\Log\LogLevel::WARNING]['TYPO3\\CMS\\Core\\Log\\Writer\\FileWriter'];
unset($GLOBALS['TYPO3_CONF_VARS']['LOG']['writerConfiguration'][\TYPO3\CMS\Core\Log\LogLevel::WARNING]['TYPO3\\CMS\\Core\\Log\\Writer\\FileWriter']);
```

**Comment by Batman777** [comment]
Awesome .. thank you. Hope this works! :-)

**Comment by Batman777** [comment]
This morning the site was online. This afternoon i got a white page.

**Log file gives me the following error:**
`PHP Catchable fatal error:  Argument 1 passed to TYPO3\\CMS\\Core\\Log\\Writer\\FileWriter::__construct() must be of the type array, null given, called in /home/XXX/typo3_src-6.2.31/typo3/sysext/core/Classes/Utility/GeneralUtility.php on line 4471 and defined in /home/XXX/typo3_src-6.2.31/typo3/sysext/core/Classes/Log/Writer/FileWriter.php on line 62`

After removing the snippet you send me yesterday the site is online again. But now the log file is growing every minute again :-(

**Comment by hannesbochmann** [comment]
Sorry, the snippet I send you was for TYPO3 7.6 upwards. I checked again in one of my 6.2 projects. That's what I put into AdditionalConfiguration.php:
```
$GLOBALS['TYPO3_CONF_VARS']['LOG']['writerConfiguration'][\TYPO3\CMS\Core\Log\LogLevel::ERROR]['TYPO3\\CMS\\Core\\Log\\Writer\\FileWriter'] =
    $GLOBALS['TYPO3_CONF_VARS']['LOG']['writerConfiguration'][\TYPO3\CMS\Core\Log\LogLevel::DEBUG]['TYPO3\\CMS\\Core\\Log\\Writer\\FileWriter'];
unset($GLOBALS['TYPO3_CONF_VARS']['LOG']['writerConfiguration'][\TYPO3\CMS\Core\Log\LogLevel::DEBUG]['TYPO3\\CMS\\Core\\Log\\Writer\\FileWriter']);
```

This works since weeks.

**Comment by Batman777** [comment]
Thank you very much again :-)

**Comment by leslawp** [comment]
Thank you guys for help, especially Hannes! I went into trouble too with huge, few GB log.

**Comment by leslawp** [comment]
Actually it resolves (in T3 7.6) problem with lot of DEBUG entries, caused by every view, but for if there is somewhere in the world strange link to our page with bad L parameter then Google will index the whole site since this bad L parameter will glue to every single link. Then lot of ERROR labeled entries will go to the log file. And voila, you have again few GB in few minutes:
This is apache log that shows how it goes:
66.249.64.207 - - [10/Nov/2017:16:07:03 +0100] "GET /index.php?id=1128&L=11 HTTP/1.1" 200 8479 "-" "Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)"
What to do? We need different approach... some kind of redirect on bad L or something like this.

**Comment by mediaessenz** [comment]
As a workaround, you can add this to your AdditionalConfiguration.php:
```
if (\TYPO3\CMS\Core\Utility\GeneralUtility::getApplicationContext()->isProduction()) {
    $GLOBALS['TYPO3_CONF_VARS'] = array_replace_recursive($GLOBALS['TYPO3_CONF_VARS'],
        [
            'LOG' => [
                'writerConfiguration' => [
                    \TYPO3\CMS\Core\Log\LogLevel::DEBUG => \TYPO3\CMS\Core\Log\Writer\NullWriter::class,
                    \TYPO3\CMS\Core\Log\LogLevel::INFO => \TYPO3\CMS\Core\Log\Writer\NullWriter::class,
                    \TYPO3\CMS\Core\Log\LogLevel::NOTICE => \TYPO3\CMS\Core\Log\Writer\NullWriter::class,
                    \TYPO3\CMS\Core\Log\LogLevel::WARNING => \TYPO3\CMS\Core\Log\Writer\NullWriter::class
                ]
            ]
        ]
    );
}
```

**Comment by leslawp** [comment]
Thank you, @mediaessenz !
I will try that!

**Comment by leslawp** [comment]
Looks like it works :)
Still we need to think about steady solution, since we are doing workarounds...

**Comment by neotrace** [comment]
I am also experiencing this problem in TYPO3 7.6. Server eventually ran out of space and crashed due to a 60GB Log file full of these RealURL Errors. I have L(0-6) set in my config since I only have those languages, but I see requests for L=10/11/12 and so on coming from outside. Probably Searchengine or bot trying languages.
Our Webhoster made a symlink to /dev/null for the log file so that is another working solution I wanted to share here. Waiting desperately for a fix.

**Comment by webian** [comment]
The huge logs are generated from here: https://github.com/dmitryd/typo3-realurl/blob/development/Classes/Encoder/UrlEncoder.php#L1589
That `debug_backtrace()` returns an array with more or less 30 elements and each of them have inside the whole `\DmitryDulepov\Realurl\Encoder\UrlEncoder` object and are later json encoded. This is repeated for every link in the page generating a ~250MB file for each page.

**Comment by webian** [comment]
I made this rewrite rule for .htaccess to remove unwanted L parameters:
RewriteCond %{QUERY_STRING} ^(.*)&?L=(?!(1|11)(&|$))[^&]+&?(.*)$
RewriteRule ^/?(.*)$ /$1?%1%4 [R=301,L]

Play with the regex: https://regex101.com/r/KsOYvR/1

In this case I allow only L=1 and L=11.
Change `1|11` with the ID that you want to allow, i.e. `2|4|5|15`, all other values will cause the L parameter to be removed.

**Comment by osahner** [comment]
#551 fixes the Logfile overflow. 
I wonder how long we have to wait for a release ...

**Comment by fazzyx** [comment]
@webian And what if u have partial translated content?

**Comment by webian** [comment]
@fazzyx sorry I can't understand what you're asking. Can you explain your doubts further?

**Comment by fazzyx** [comment]
@webian Basically you translate to another language, but not every content / site.

**Comment by webian** [comment]
@fazzyx my rewrite rule should be used to remove L parameter with values not configured. Does your case have urls with values for L parameter not configured?
ie.: you use default language English (uid=0) and a second language german (uid=1). Then you change the regex to allow only L=0 or L=1: `^(.)&?L=(?!(0|1)(&|$))[^&]+&?(.)$`.
Did I miss something?

**Comment by fazzyx** [comment]
@webian I just described a scenario where not every content or page are translated. You have only a workaround more, but we need a solution.

**Comment by franzkugelmann** [comment]
we were able to stop the logging with TypoScript
`config.linkVars := addToList(L(0|1|4))`
You need to add your language-ids instead of 0|1|4.
Then TYPO3 does not let the not-existing language id through and the logger in realurl is not triggered.

**Comment by LeoniePhiline** [comment]
+1. We are also badly affected by this bug, with debug_backtrace() filling up Gigabytes of logs until the server hard drive runs out and the site shuts down.
This should seriously be fixed. debug_backtrace() is not even helping.

**Comment by fazzyx** [comment]
@franzkugelmann 
As I understand this, It is only working, if every page is translated but not if you have a partially translated environment.
If you have setup translation to a foreign language, but not every page is translated to this language you need the linkVars id for this language anyway
and therefor it is possible to call a page which is not translated with one of the defined linkVars id's which in this case is invalid for this page.

**Comment by franzkugelmann** [comment]
@fazzyx 
Yes, you are right. `config.linkVars` only helps with non-existing languages (as defined in table sys_language), but not with just-not-translated pages. Our problems occured due to a deleted language which was not supported by the editors anymore.

**Comment by Bjoelin** [comment]
I am currently playing around with the configuration and just realized; this als ooccurs with "correct" L parameters. 

I my case, 3 is set as sys_language_uid for Dutch content. Setting config.defaultGetVars.L = 3 breaks url rewriting and produces huge logs as well.

`Bad "L" parameter ("3") was detected by realurl.`

Can anyone confirm this problem?

**Comment by dmitryd** [comment]
Please, do not use debug logging in production. That's all.

**Comment by dmitryd** [comment]
@Bjoelin This warning is produced if you have no language record with uid=3 in your system.

**Comment by franzkugelmann** [comment]
A hint about the production context dmitry mentioned:
If you wonder why you get the big dev context logging on your production site, check your devIpMask! 
See the description in the install tool at [SYS][debugExceptionHandler]
We had the customers IP in there and thought this perfectly valid. Until they started running a SEO crawler, which resulted in thousands of requests triggering the debugExceptionHandler.

**Comment by auge** [comment]
for YPO3-8.7 I had to pass an array instead of the classname:

`$GLOBALS['TYPO3_CONF_VARS']['LOG']['writerConfiguration'][\TYPO3\CMS\Core\Log\LogLevel::WARNING]['TYPO3\\CMS\\Core\\Log\\Writer\\FileWriter'] = array(\TYPO3\CMS\Core\Log\Writer\NullWriter::class => array())`

the key of the array is the classname to be factored, the value are options (just empty array here).

**Comment by pniederlag** [comment]
This is really an issue as an invalid L parameter is not that unlikely to happen.

Removing/Unsetting the Log or setting it to the NullWriter does work of course but actually is more of a workarund.

@dmitryd could you please remove the debug_backtrace() from the error message? It really quickly fills the disk with hundreds of GB logs! (we have ~250MB per single request with a broken L parameter).

My suggestions/pledge is:
- lower priority to warning
- add some useful information but not a full backtrace

If you are willing to accept a PR I'd be happy to prepare one.

nonetheless, thx for your great work,
Peter

**Comment by pniederlag** [comment]
@dmitryd 

> Please, do not use debug logging in production. That's all.

The problem does occur/happen with production exceptionhandlers and all other relevant settings set for a production context.

`$this->logger->error($message,` debug_backtrace());

In general you'd want to catch errors in production settings, however you'd not want it to contain result of a call to debug_backtrace() ;)

**Comment by LeoniePhiline** [comment]
@pniederlag +1

**Comment by nabossha** [comment]
+1 at least in production-context it should be reduced to a simple error-message

**Comment by argerweb** [comment]
@pniederlag +1

**Comment by argerweb** [comment]
@pniederlag +1

**Comment by mhelwig** [comment]
This effectively means that any simple security scanner can shoot down your Typo3 page in a normal server environment within minutes. It's a severe DoS vulnerability. One of my servers generated approx. 450 MB Iogs for each invalid request. I also reported this to the Typo3 security team a while ago but since it's already public they don't seem to care.

**Comment by dmitryd** [comment]
Should be fixed now.

**Comment by Outdoorsman** [comment]
I'll check out that fix... thanks for the @dmitryd. This took down a trove of our websites in a shared environment because of a single massive log file.

**Comment by Batman777** [comment]
So has RealURL 2.3.1 fixed that  issue ?

**Comment by Outdoorsman** [comment]
@Batman777 I'm still testing to see if `2.3.1` works for me. I'm surprised that I'm having an issue with this at all because I only have one language set. The only place I've found reference other languages is conditionals within the `bootstrap_package`, could this be triggering the additional L parameters?

```
[globalVar = GP:L = 1]
    config {
        sys_language_uid = 1
        language = de
        locale_all = de_DE.UTF-8
        htmlTag_setParams = lang="de" dir="ltr" class="no-js"
    }
[global]
[globalVar = GP:L = 2]
    config {
        sys_language_uid = 2
        language = da
        locale_all = da_DK.UTF-8
        htmlTag_setParams = lang="da" dir="ltr" class="no-js"
    }
[global]
```

**Comment by dmitryd** [comment]
@Outdoorsman

Check if your config has `config.linkVars = L(0-2)` (it should!). Also check for `addQueryString` in TypoScript and viewHelpers.

Also read this: https://github.com/dmitryd/typo3-realurl/wiki/Notes-for-Integrators#i-see-a-lot-of-strange-entries-in-the-realurl-module

**Comment by adrienjacob** [comment]
There's a typical issue with bootstrap_package : you need to disable unneeded languages in the template constants - `page.theme.language.languageValue = 0` (if you only use the default language)

cf. this other thread :
https://github.com/dmitryd/typo3-realurl/issues/407#issuecomment-285704553

**Comment by wtfred** [comment]
@Outdoorsman Did you get the 2.3.1 from TYPO3 TER or directly from github ? I made the update from the TER but i'm still getting huge log files...

**Comment by sgrossberndt** [comment]
2.3.1 does not contain the fix: https://github.com/dmitryd/typo3-realurl/commit/2c669925ebdec3212d01996b79507bdaba29db11#commitcomment-26436691

**Comment by Outdoorsman** [comment]
When I first updated it, it seemed to keep spewing log files that went in the the gigabytes within minutes. But if I remember right (and I'm not sure about this), I cleared cache from the TYPO3 backend and it didn't work but then cleared them from the Install Tool which I believe does it much more thoroughly and then it reduced down to single line entries instead of a full trace... at least manageable if you don't leave it too long. I used the Development branch from... https://github.com/dmitryd/typo3-realurl.git.

**Comment by franzholz** [comment]
I do not understand the wiki page 
  https://github.com/dmitryd/typo3-realurl/wiki/Notes-for-Integrators#i-see-a-lot-of-strange-entries-in-the-realurl-module
Why are these strange L parameters generated from anywhere inside of TYPO3? I thought all links are generated by realurl if activated. Therefore these links with wrong L parameters should not exist at all. 
 
    [08-Jan-2018 19:27:02 Europe/Berlin] http://mywebsite.com/ - cms: $TSFE->set_no_cache() was triggered. Reason: Bad "L" parameter ("courier-authlib-0.59.1-1.x86_64.rpm") was detected by realurl. Page caching is disabled to prevent spreading of wrong "L" value.. Caching is disabled!

If I use the TYPO3 backend to search for a term "courier-authlib-0.59.1-1.x86_64", then I have no result.

**Comment by Outdoorsman** [comment]
> There's a typical issue with bootstrap_package : you need to disable unneeded languages in the template constants - `page.theme.language.languageValue = 0` (if you only use the default language)

@adrienjacob Thanks for that tip about the above issue in bootstrap_package. **Unfortunately the fix you mentioned only works for [version 7.1.0](https://github.com/benjaminkott/bootstrap_package/tree/7.1.0) and above and I have [bootstrap_package version 7.0.5](https://github.com/benjaminkott/bootstrap_package/tree/7.0.5).** It's also in a complicated integration so I can't just update it quickly.

**Comment by dmitryd** [comment]
@franzholz 
> I thought all links are generated by realurl if activated. Therefore these links with wrong L parameters should not exist at all.

No. Links are generated to TYPO3 and only passed to realurl to encode to speaking url. So if TYPO3 makes `/index.php?id=123&L=test`, realurl can only complain that it got invalid `L`.

**Comment by jean-alex** [comment]
The fix suggested by @adrienjacob solved my problem.

**Comment by daCyberpunk** [comment]
I'm using also the workaround mentioned [here](https://github.com/dmitryd/typo3-realurl/issues/549#issuecomment-343657476), but I'm not happy with. Because on our staging system we have the same problem and if we are using the workaround, then we can not set the  Debug settings in Installtool to debug.
I also tried to set `page.theme.language.languageValue = 0,1,6,9,10` and i tried `config.linkVars = L(0-10)` as well as `config.linkVars = :=addToList(L(0|1|6|9|10))` but still we get the error about the bad L parameter (2). 
What I can add to this discussion is, that we have in Backend added the french language which actually has the id 2 but we set it to hidden.  I don't know if that could be the problem.

**Comment by franzholz** [comment]
@Dmitry: Now I did more investigations. I collected the IP addresses of the users who's visits to my website generate those masses of entries into the PHP error_log file by realurl:

    [24-Jan-2018 09:29:30 Europe/Berlin] http://www.mywebsite.com/ - cms: $TSFE->set_no_cache() was triggered. Reason: Bad "L" parameter ("links") was detected by realurl for IP 46.229.168.73. Page caching is disabled to prevent spreading of wrong "L" value.. Caching is disabled!
    [24-Jan-2018 09:29:30 Europe/Berlin] http://www.mywebsite.com/ - cms: $TSFE->set_no_cache() was triggered. Reason: Bad "L" parameter ("links") was detected by realurl for IP 46.229.168.73. Page caching is disabled to prevent spreading of wrong "L" value.. Caching is disabled!
    [24-Jan-2018 09:29:48 Europe/Berlin] http://mywebsite.com/ - cms: $TSFE->set_no_cache() was triggered. Reason: Bad "L" parameter ("seite-nicht-gefunden") was detected by realurl for IP 216.244.66.200. Page caching is disabled to prevent spreading of wrong "L" value.. Caching is disabled!
    [24-Jan-2018 09:29:48 Europe/Berlin] http://mywebsite.com/ - cms: $TSFE->set_no_cache() was triggered. Reason: Bad "L" parameter ("seite-nicht-gefunden") was detected by realurl for IP 216.244.66.200. Page caching is disabled to prevent spreading of wrong "L" value.. Caching is disabled!
    [24-Jan-2018 09:29:48 Europe/Berlin] http://mywebsite.com/ - cms: $TSFE->set_no_cache() was triggered. Reason: Bad "L" parameter ("seite-nicht-gefunden") was detected by realurl for IP 216.244.66.200. Page caching is disabled to prevent spreading of wrong "L" value.. Caching is disabled!

Both IP addresses are listed as known for making attempts for DDoS attacks. 

Do you still continue to think that TYPO3 generated those wrong L parameters?

**Comment by dmitry** [comment]
@dmitryd

**Comment by co-operate** [comment]
Same behaviour here on several sites with TYPO3 8.7.9 + RealURL 2.3.1. Most notably on one site with `basic.enableAutoConf=true` and `config.linkVars=L(0-1)` set and in standard production envirnoment settings (including the standard `'SYS/systemLogLevel' => '2'`). On this site, the log file entries seem to be triggered by sql injection attempts:

`Sun, 28 Jan 2018 07:48:40 +0100 [ERROR] request="6fac06f4c348d" component="DmitryDulepov.Realurl.Encoder.UrlEncoder": Bad "L" parameter ("1 AND 1=2 UNION SELECT 0x6461726b31636f6465--\'A=0\'A=0") was detected by realurl. Page caching is disabled to prevent spreading of wrong "L" value`

As a consequence, the log file currently grows by 2 GB/h. Each new line adding >20MB to the log file. I think that's not acceptable in a production environment...

**Comment by DiablosOffens** [comment]
I had the same impression in the past, but found the actual root cause of the problem on our website.
It wasn't a problem in typo3 core or the realurl extension. We use an [extension](https://extensions.typo3.org/extension/forcerealurls2/) which forces all requesters to redirect to URLs provided by realurl if the requested URL includes "index.php". I fixed the bug there and now all log messages are gone.
I provided this fix for the author's [git repository](https://bitbucket.org/edirect24ug/forcerealurl2).

The stack trace in the log messages were actually a real help to find this bug.

**Comment by Outdoorsman** [comment]
> I had the same impression in the past, but found the actual root cause of the problem on our website.
It wasn't a problem in typo3 core or the realurl extension. We use an extension which forces all requesters to redirect to URLs provided by realurl if the requested URL includes "index.php". I fixed the bug there and now all log messages are gone.
I provided this fix for the author's git repository.

Would it make sense for something like the `forcerealurls2` to be integrated into realurl out of the gate? I see this extension still doesn't support TYPO3 8.7, which is a deal breaker for some of my sites.

**Comment by DiablosOffens** [comment]
> Would it make sense for something like the forcerealurls2 to be integrated into realurl out of the gate? I see this extension still doesn't support TYPO3 8.7, which is a deal breaker for some of my sites.

Yeah sure, removing the need for such outdated extensions would also reduce the possible attack vectors.
But it has to be something like the fixed version I'm currently using.

**Comment by CDRO** [comment]
+1

**Comment by ribase** [comment]
Is it possible to lockdown 2.3.1 and 2.3.0? At this point we have to set a fix version in every composer.json...

This is a serious bug and will "DDOS" the system itself...

**Comment by Outdoorsman** [comment]
> Is it possible to lockdown 2.3.1 and 2.3.0? At this point we have to set a fix version in every composer.json... This is a serious bug and will "DDOS" the system itself...

This has taken down an entire server full of websites multiple times. After some work I have been able to reduce the amount of logs but they are still filling up with imaginary errors as far as I can tell. The default should not be to blow up the system with logs like this, maybe you can add a debug checkbox mode to the extension or something but when in production this shouldn't happen!

**Comment by xperseguers** [comment]
Same problem here, logs are filling up my server disk constantly and very rapidly. Version 2.3.1 installed.

**Comment by No5251** [comment]
If you're using the "bootstrap_package", set correct values for
`config.linkVars` and `page.theme.language.languageValue`.

**Comment by dmitryd** [comment]
@Outdoorsman 

> This has taken down an entire server full of websites multiple times. After some work I have been able to reduce the amount of logs but they are still filling up with imaginary errors as far as I can tell. The default should not be to blow up the system with logs like this, maybe you can add a debug checkbox mode to the extension or something but when in production this shouldn't happen!

May be, you can configure your system correctly after all this? What do you think? ;) These are not imaginary errors but a faulty implementation of your site.

**Comment by dmitryd** [comment]
@co-operate Same for you: configure the site correctly! Your site allows arbitrary `L` values. The one from logs shows that you allow values that contain SQL injections on your site.

**Comment by wtfred** [comment]
> Same for you: configure the site correctly! Your site allows arbitrary L values. The one from logs shows that you allow values that contain SQL injections on your site.

> Same behaviour here on several sites with TYPO3 8.7.9 + RealURL 2.3.1. Most notably on one site with basic.enableAutoConf=true and config.linkVars=L(0-1) set and in standard production envirnoment settings (including the standard 'SYS/systemLogLevel' => '2'). On this site, the log file entries seem to be triggered by sql injection attempts:
`Sun, 28 Jan 2018 07:48:40 +0100 [ERROR] request="6fac06f4c348d" component="DmitryDulepov.Realurl.Encoder.UrlEncoder": Bad "L" parameter ("1 AND 1=2 UNION SELECT 0x6461726b31636f6465--\'A=0\'A=0") was detected by realurl. Page caching is disabled to prevent spreading of wrong "L" value`

@dmitryd I thought `config.linkVars=L(0-1)` was the way to restrict `L` values to 0 and 1 only ? What's wrong with his configuration ?

**Comment by Outdoorsman** [comment]
@dmitry as far as I can tell I've followed your advice on how to configure and still get errors. Don't know what to think, but I still say that I've used this extension for years and never had an issue until now, all of a sudden, things are blowing up. I didn't change anything on my end. Then I tried to follow the directions you gave and it though it did reduce things I still couldn't get it to stop slowing filling my logs. I love verbose logging when needed but I still think the default should not be something so verbose that it crashes servers. Please don't take it personally... I'm just sharing one opinion of what would be good practice. I will still use and be grateful for the extension either way.

**Comment by fazzyx** [comment]
If I change the page content, maybe you drop a (multilanguage) page (but not the whole language), there are still entries anywhere outside
and if a bot calling these sites the log file get flooded.
How can I managed this, what is the right setup for this?

**Comment by dmitryd** [comment]
@fazzyx @Outdoorsman Read: https://github.com/dmitryd/typo3-realurl/wiki/Notes-for-Integrators#i-see-a-lot-of-strange-entries-in-the-realurl-module This gives instructions what to do.

**Comment by Anubiso** [comment]
I have a single-language page and also get log entries. My Config looks like this:

**TypoScript-Setup:**
 ```
config{
    linkVars = L(0)
    sys_language_uid = 0
    language = de
}
```

**realurlconf:** 
```
'preVars' => array(
    array(
        'GETvar' => 'L',
        'noMatch' => 'bypass',
    ),
)
```

TYPO3 doesn't generate the faulty links, but when i call the page with a faulty L parameter manually the log file is flooded. Worst case: Using a faulty L parameter on a paginated news list page where 25 log entries (250MB) are created on 1 request.

**Example-URL valid:** 
```
index.php?id=113&tx_news_pi1%255B%2540widget_0%255D%255BcurrentPage%255D=2&cHash=ea27ba24a8da3abdbb255e329207e874
```

**Flood the logs:**
```
id=113&L=142&tx_news_pi1%255B%2540widget_0%255D%255BcurrentPage%255D=2&cHash=ea27ba24a8da3abdbb255e329207e874
```

**Comment by Anubiso** [comment]
Additional Info:
I think the problem is a core problem as the UriBuilder used in the widget-link view-helper obviously doesn't validate the language parameter. (the wrong L parameter is used in all pagination links) 
`<f:widget.link` addQueryStringMethod="GET"`

Nevertheless i think there should be an option to disable/shorten the log made by realurl, because there is no easy way to fix this.

**Comment by dmitryd** [comment]
@Anubiso 

> TYPO3 doesn't generate the faulty links, but when i call the page with a faulty L parameter manually the log file is flooded.

Contradiction here. If log is flooded, it means that TYPO3 called realurl with a wrong `L` value. You can direct logs to nowhere.
